weave/integrations/langchain/helpers.py

```diff
@@ -0,0 +1,272 @@
+    """
+    prompt_tokens = 0
+    completion_tokens = 0
+    total_tokens = 0
+
+    # OpenAI format: prompt_tokens, completion_tokens, total_tokens
+    if "prompt_tokens" in usage_data and "completion_tokens" in usage_data:
+        prompt_tokens = usage_data.get("prompt_tokens", 0) or 0
+        completion_tokens = usage_data.get("completion_tokens", 0) or 0
+        total_tokens = usage_data.get("total_tokens", 0) or 0
+
+    # Google GenAI format: input_tokens, output_tokens, total_tokens
+    elif "input_tokens" in usage_data and "output_tokens" in usage_data:
+        prompt_tokens = usage_data.get("input_tokens", 0) or 0
+        completion_tokens = usage_data.get("output_tokens", 0) or 0
+        total_tokens = usage_data.get("total_tokens", 0) or 0
+
+    # Google Vertex AI format: prompt_token_count, candidates_token_count, total_token_count
+    elif "prompt_token_count" in usage_data and "candidates_token_count" in usage_data:
+        prompt_tokens = usage_data.get("prompt_token_count", 0) or 0
+        completion_tokens = usage_data.get("candidates_token_count", 0) or 0
+        total_tokens = usage_data.get("total_token_count", 0) or 0
+
+    return prompt_tokens, completion_tokens, total_tokens
+
+
+def _extract_model_from_flattened_path(flattened: dict, usage_key_path: str) -> str:
+    """
+    Extract model name from flattened response structure.
+
+    Searches for model_name in common locations relative to usage metadata path.
+    Falls back to broader search if targeted patterns fail.
+
+    Args:
+        flattened: Flattened response with dot-separated keys
+        usage_key_path: Key path where usage metadata was found
+
+    Returns:
+        Model name or "unknown" if not found
+
+    Common patterns:
+        usage at "outputs.0.usage_metadata" → model at "outputs.0.generation_info.model_name"
+    """
+    path_parts = usage_key_path.split(".")
+
+    # Generate specific candidate paths based on observed provider patterns.
+    # We construct paths by replacing the usage-related components with known
+    # model name locations. This handles the most common scenarios where model
+    # names appear in predictable relative locations.
+    search_patterns = []
+
+    # Generate patterns by truncating at different levels and adding model paths
+    for truncate_count in range(1, min(len(path_parts), 6)):
+        base_path = path_parts[:-truncate_count]
+        # generation_info patterns
+        search_patterns.append(".".join(base_path + ["generation_info", "model_name"]))
+        # response_metadata patterns
+        search_patterns.append(
+            ".".join(base_path + ["response_metadata", "model_name"])
+        )
+
+    for pattern in search_patterns:
+        if pattern in flattened:
+            model_name = flattened[pattern]
+            if isinstance(model_name, str) and model_name:
+                return model_name
```

Comment:
what is going on here?

Reply:
how brittle is this?  I don't really understand what `range(1, min(len(path_parts), 6))` is supposed to mean -- did you just make that up?

Reply:
6 was based on the observed dictionaries is among most langchain providers. I chose it to balance coverage and performance, assuming most provider response paths won’t be deeper than 6 segments from the usage metadata. But you are right, it could be brittle and break things. So. I'll change to do a full check instead.